mlora/executor/task/dpo_task.py

```diff
@@ -0,0 +1,164 @@
+            "ipo": self.__dpo_loss_ipo
+        }
+
+        self.context_.set_loss_fn(LOSS_CLASS[self.config_.loss_type_])
+
+    def _pre_ref_context(self, linears_info: OrderedDict[str, LinearInfo]):
+        if self.config_.reference_ is None:
+            self.ref_context_ = None
+            return
+
+        self.ref_context_ = TASKCONTEXT_CLASS[self.config_.reference_.type_](
+            self.config_.reference_.name_)
+        self.ref_context_.init(self.config_.reference_, linears_info)
+
+    def __dpo_loss_sigmoid(self, logits: torch.Tensor) -> torch.Tensor:
+        loss = -F.logsigmoid(self.config_.beta_ * logits) * \
+            (1 - self.config_.label_smoothing_) - \
+            F.logsigmoid(-self.config_.beta_ * logits) * \
+            self.config_.label_smoothing_
+        return loss
+
+    def __dpo_loss_ipo(self, logits: torch.Tensor) -> torch.Tensor:
+        loss = (logits - 1 / (2 * self.config_.beta_)) ** 2
+        return loss
+
+    def adapter_model(self) -> List[AdapterModel]:
+        if self.ref_context_ is None:
+            return super().adapter_model()
+
+        return [self.context_.adapter_model(), self.ref_context_.adapter_model()]
+
+    def adapter_name(self) -> List[str]:
+        if self.ref_context_ is None:
+            return super().adapter_name()
+
+        return [self.config_.adapter_.name_, self.config_.reference_.name_]
+
+    def switch_device(self, device: str):
+        if self.ref_context_ is not None:
+            self.ref_context_.switch_device(device)
+
+        self.context_.switch_device(device)
+
+    def data(self, start_idx: int) -> Tuple[List[Tokens], List[MLoRADataConfig]]:
+        logging.info(
+            f'Task - {self.context_.name_} epoch: {self.now_epoch_}/{self.config_.num_epochs_}'
+            f' iteration: {self.now_data_idx_}/{len(self.data_)} step: {self.now_step_}')
+        data_idx_s = self.now_data_idx_
+        data_idx_e = self.now_data_idx_ + self.config_.mini_batch_size_
+
+        # 0...mid is chosen data
+        # mid.end is reject data
+        batch_str = self.prompter_.generate_prompt_batch(
+            self.data_[data_idx_s:data_idx_e])
+
+        assert len(batch_str) % 2 == 0
+
+        ret_tokens = []
+        # for refrerence
+        ref_model_token = list(map(lambda raw_str: self.tokenizer_.encode(
+            raw_str, bos=True, eos=True), batch_str))
+        policy_model_token = copy.deepcopy(ref_model_token)
+
+        ret_tokens.extend(ref_model_token)
+        ret_tokens.extend(policy_model_token)
+
+        # include reference and policy models' chosen and reject data
+        assert len(ret_tokens) % 4 == 0
+
+        ref_start_idx = start_idx
+        ref_end_idx = ref_start_idx + len(ref_model_token)
+
+        policy_start_idx = ref_end_idx
+        policy_end_idx = policy_start_idx + len(policy_model_token)
+
+        def loss_fn(input: torch.Tensor, target: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
```

Comment:
can you provide the related forumal for this loss function

Reply:
yes, we will update in document.